################################ General Parameters #############################################
general:
  # Specify the mode of operation for the segmentation pipeline
  # 0: Semantic Segmentation runs in conjunction with Geometric Segmentation
  # 1: Semantic Segmentation runs independently
  # 2: Geometric Segmentation runs independently
  # run with 2 if semantic segmentation is not running - this clears the pointclouds
  mode_of_operation: 1

  # The path of the high level semantic information of the environment
  # [TODO] - the path looks ugly, possible workaround?
  env_database: "../../../src/visual_sgraphs/config/Environments/LuxVSD/luxvsd_seq01.json"

#################################### Marker Parameters ###########################################
markers:
  # How much the framework should trust the markers (0.1 default, higher for mono setups)
  impact: 0.1

################################ Common Segmentation Parameters ##################################
seg:
  # Number of points (after filtering) needed to detect geometric objects (e.g., planes)
  # Also, the number of points needed to detect a plane in the pointcloud
  pointclouds_thresh: 400

  # Plane association threshold (ominus)
  plane_association_thresh: 0.20

  # The maximum distance for point to be considered part of a plane
  plane_point_dist_thresh: 0.01

  ransac:
    # The maximum number of planes to be detected in one pointcloud
    max_planes: 2

    # The maximum distance for a point to be considered an inlier of the plane model
    distance_thresh: 0.01

    # The maximum number of iterations for the RANSAC algorithm
    max_iterations: 800

################################## Optimization Parameters #######################################
optimization:
  # plane equation to keyframe factor
  plane_kf:
    enabled: true
    information_gain: 1.5
  # point to plane factor
  plane_point:
    enabled: false
    information_gain: 0.002

############################ Geometric Segmentation Parameters ####################################
geo_seg:
  pointcloud:
    # pointcloud downsampling parmeters
    downsample:
      leaf_size: 0.05
      min_points_per_voxel: 25
    # Pointcloud outlier removal
    outlier_removal:
      mean_threshold: 50
      std_threshold: 1.0

############################## Semantic Segmentation Parameters ####################################
sem_seg:
  pointcloud:
    # pointcloud downsampling parmeters
    downsample:
      leaf_size: 0.03
      min_points_per_voxel: 25
    # Pointcloud outlier removal
    outlier_removal:
      mean_threshold: 50
      std_threshold: 1.0

  # The threshold for semantic segmentation probability validation (rejects probability below this)
  prob_thresh: 0.5

  # The threshold for semantic segmentation confidence validation (rejects confidence below this)
  conf_thresh: 0.3

  # The maximum height above the main ground to be a valid stepped ground plane (in meters)
  max_step_elevation: 0.5

  # The maximum tilt heuristic for a wall/ground plane (checks respective component of plane equation)
  max_tilt_wall: 0.2
  max_tilt_ground: 0.2

  # The amount of votes (votes are weighted votes) for a plane to be finally classified with a semantic label
  min_votes: 2.5

  # Whether to reassociate planes with the semantic labels - after optimization, some planes may get close
  reassociate: 
    enabled: false
    association_thresh: 0.1

  ############################## Room Segmentation Parameters ####################################
room_seg:
  # The method to use for room segmentation
  # 0: Segment geometrically based on closest walls
  # 1: Segment based on free-space clustering (skeleton voxblox)
  # 2: Segment based on GNN
  method: 1

  # The maximum threshold for the dot product of the plane normals to be considered facing each other
  plane_facing_dot_thresh: -0.5

  # Threshold for perpendicularity of wall pairs to be considered a square room (in degrees)
  perpendicularity_thresh: 10.0

  # The maximum distance between two room centroids to be considered the same room (in meters)
  room_center_distance_thresh: 1.0

  # Parameters for the geometrically-based closest walls room segmentation
  geo_based:
    # The maximum distance from a marker to a wall to be considered part of the room
    marker_wall_distance_thresh: 6.0

  # Parameters for skeleton voxblox free-space room segmentation
  skeleton_based:
    # The minimum number of points needed to form a cluster
    min_cluster_vertices: 5
    # The maximum distance from a point of a cluster to a wall to be considered part of the room
    cluster_point_wall_distance_thresh: 0.3
